src/js/AuditRules.js

```diff
@@ -16,21 +16,46 @@ goog.require('axs.AuditRule');
 
 goog.provide('axs.AuditRules');
 
-/**
- * Gets the audit rule with the given name.
- * @param {string} name
- * @return {axs.AuditRule}
- */
-axs.AuditRules.getRule = function(name) {
-    if (!axs.AuditRules.rules) {
-        /** @type Object.<string, axs.AuditRule> */
-        axs.AuditRules.rules = {};
-        for (var specName in axs.AuditRule.specs) {
-            var spec = axs.AuditRule.specs[specName];
-            var auditRule = new axs.AuditRule(spec);
-            axs.AuditRules.rules[spec.name] = auditRule;
-        }
-    }
+(function(){
+    var auditRulesByName = {};
+    var auditRulesByCode = {};
 
-    return axs.AuditRules.rules[name];
-};
+    /**
+     * Instantiates and registers an audit rule.
+     * If a conflicting rule is already registered then the new rule will not be added.
+     * @param {axs.AuditRule.Spec} spec The object which defines the AuditRule to add.
+     * @throws {Error} If the rule duplicates properties that must be unique.
+     */
+    axs.AuditRules.addRule = function(spec) {
+        // axs.AuditRule.specs[spec.name] = spec;  // This would add backwards compatibility
+        // create the auditRule before checking props as we can expect the constructor to perform the
+        // first layer of sanity checking.
+        var auditRule = new axs.AuditRule(spec);
+        if (auditRule.code in auditRulesByCode || auditRule.name in auditRulesByName)
```

Comment:
How about splitting these out into two separate cases so we can have two separate error messages?